### ISSUE TITLE
Preferred way to commit/rollback 

### ISSUE THREAD

**Issue by artslob** [tiangolo/fastapi#2142]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

---
When read about FastAPI's dependency system with `yield` state, its pretty tempting to use database transactions with them like so:
```
async def get_db(request: Request):
    """ database: Database = Depends(get_db) """
    database: Database = request.app.state.database
    async with database.transaction():
        yield database
```
But this is incorrect way to do it because if transaction would not commit (teardown code in `async with` raises error) then there is no way to change response.

So case about transactions would be good to note in docs for dependencies with `yield` statement. Maybe add link to starlette docs about transactions: https://www.starlette.io/database/#transactions.

**Comment by sm-Fifteen** [comment]
As per [PEP 249](https://www.python.org/dev/peps/pep-0249/), connections aren't supposed to auto-commit changes. Any work that hasn't been explicitly committed before you return and your dependency closes the transaction will be rolled back.

>  Note that closing a connection without committing the changes first will cause an implicit rollback to be performed.

It wouldn't make sense to do it the other way around, because that would mean that any uncaught exception would commit the exception, leaving the database in an unclean state. I don't know what SQL toolkit you're using, but unless it's doing some magic in regards to committing if the context manager exits normally and rolling back on error, you shouldn't be getting this sort of issue.

**Comment by artslob** [comment]
I use [encode/databases](https://www.encode.io/databases/) as SQL toolkit. And it actually commits all changes when used without transaction.

>committing if the context manager exits normally and rolling back on error

Yes, its working this way. Basically I want my every endpoint to be executed inside transaction (even just simple reads). And I want it to be as automatic as possible, without explicitly opening transaction in every enpoint. To achieve this I used dependency.

The problem is that using dependency system for transaction management (I provided example with `yield`) is wrong and should be noted in docs (because commit can fail and there is no way to change response in teardown code of dependency). But many people may want to do transactions this way and we should warn them. Also it would be good to provide some best practicies how to manage transactions in FastAPI without using dependency injection system, because any production code would be using transactions with high chance.

Right now I decided to use explicit transactions with `async with` block in every endpoint.

**Comment by sm-Fifteen** [comment]
Ah, right, I believe encode/databases doesn't use transactions by default. I don't personally believe that transactions are a good fit for context-manager dependencies, given the scope for which dependencies are kept open extends all the way until the response has been sent and any [background tasks](https://fastapi.tiangolo.com/tutorial/background-tasks/) have finished running, so you can't actually rely on their exit status, which you have to because a transaction failing to commit means your operation went wrong and the 200/201 status is then incorrect.

If you want all DB interactions to be wrapped in transactions (which is a fair thing to want), you'd have better luck looking at what configuration options databases has in that regard (if any, the project is definitely not as mature as other encode projects). Implicit autocommit is something that makes sense if you want your toolkit to mirror how things would work in an interactive SQL editor, but is rarely recommended in applications and [SQLAlchemy is planning to remove support for autocommit in the 2.0 release](https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#library-level-but-not-driver-level-autocommit-removed-from-both-core-and-orm). The consensus in most of the Python SQL ecosystem seems to be that it shouldn't be your responsibility to wrap your DB operations into a transaction unless you want finer control over what parts of your operation are considered atomic, and in those cases your transaction context manager belongs in your route function, not in a dependency.

**Comment by artslob** [comment]
Good. I vote for 2 things:
1. Add a warning to dependency injection docs that its bad place to make a transaction
2. Maybe add link to Starlette docs about transactions

**Comment by lephuongbg** [comment]
You could use middleware to have every endpoint automatically wrapped in transaction like us:

```python
async def automatic_transaction(request: Request, call_next):
    """Wrap all data-modifying requests in transaction"""
    from models import db # We use peewee

    if request.method.lower() in ("post", "put", "patch", "delete"):
        with db.atomic(): # Context manager will auto-rollback transaction on exception
            response = await call_next(request)
    else:
        response = await call_next(request)

    return response

app.add_middleware(BaseHTTPMiddleware, dispatch=automatic_transaction)
```

**Comment by artslob** [comment]
@lephuongbg Nope, it would not work. Try to raise `HTTPException` in handler. Transaction will commit successfully, because `HTTPException` is [handled](https://www.starlette.io/exceptions/#errors-and-handled-exceptions) exception, so exception context would not be raised nor in middlewares nor in dependencies and transaction context manager would not see any exception:
> Handled exceptions do not represent error cases. They are coerced into appropriate HTTP responses, which are then sent through the standard middleware stack. By default the HTTPException class is used to manage any handled exceptions.
>
> Errors are any other exception that occurs within the application. These cases should bubble through the entire middleware stack as exceptions. Any error logging middleware should ensure that it re-raises the exception all the way up to the server.

**Comment by lephuongbg** [comment]
@artslob Please check the updated script for handling in case of `HTTPException`.

**Comment by artslob** [comment]
@lephuongbg Yeah, it can work. But it requires one more `try..except` because transaction's commit can fail when there is 200 status code.

I also think thats automatic transactions can be achieved by analyzing `APIRoute` instances. They have `endpoint` function that can be wrapped with transaction:
```python
def create_app() -> FastAPI:
    app = FastAPI()

    app.include_router(routers.common.router, prefix="/common", tags=["<common>"])

    install_transaction_decorator(app, database)

    return app


def install_transaction_decorator(app: FastAPI, database: Database):
    routes: List[APIRoute] = [r for r in app.routes if isinstance(r, APIRoute)]
    for r in routes:
        # here you can introspect methods, dependencies, etc of this endpoint 
        r.endpoint = database.transaction_wrapper(r.endpoint)
```

**Comment by gocreating** [comment]
Any conclusion here? I am also confused about how to manage a transaction in fastapi.

Is fastapi possible to fulfill these requirements?
1. no decorator and middleware
2. dependency only
3. commit after path operation successfully executed but before sending response
4. rollback if path operation raises any exception

**Comment by artslob** [comment]
@gocreating
If you create transaction in dependency and:
1. use handled exceptions - its not possible (because dependency would not catch exception)
2. use **un**handled exceptions - dependency would catch exception, there would be rollback, but client always will get 500 http code and one hardcoded error message (FastAPIs standart or yours). Its not convenient, because you want give 404 code to your client when item not found, 400 code when request is invalid etc.

IMHO best way is to wrap every path function to transaction wrapper or use transactions in function explicitly, like so:
```python
async def endpoint(db: Database):
    async with db.transaction():
        return await database.execute("sql");
```

**Comment by gocreating** [comment]
@artslob Thanks for the feedback.
I've proposed a decorator to fulfill my personal requirement:
https://github.com/tiangolo/fastapi/issues/2662

**Comment by andredias** [comment]
Since almost all endpoints need a transaction, I believe the best way to provide it to them is through middleware rather than decorators. Using [encode/databases](https://www.encode.io/databases/), a good solution is:

```python

from databases import Database
from fastapi import FastAPI, Request, Response
from starlette.middleware.base import BaseHTTPMiddleware

db = Database('sqlite:///')
app = FastAPI()

@app.on_event('startup')
async def startup_event():
    await db.connect()

@app.on_event('shutdown')
async def shutdown_event():
    await db.disconnect()


async def db_transaction_wrapper(request: Request, call_next) -> Response:
    async with db.connection() as conn:  # uses the current connection
        async with conn.transaction():
            return call_next(request)


app.add_middleware(BaseHTTPMiddleware, dispatch=db_transaction_wrapper)
```

You can see a more complete solution in my project, [cookiecutter-fastapi](https://github.com/andredias/cookiecutter-fastapi).
